vllm/core/block_manager_v2.py

```diff
@@ -227,20 +228,69 @@ def fork(self, parent_seq: Sequence, child_seq: Sequence) -> None:
-                num_lookahead_slots: int) -> Dict[int, int]:
-        raise NotImplementedError
+        return self._can_swap(seq_group, Device.GPU, SequenceStatus.SWAPPED,
+                              num_lookahead_slots, self.watermark_blocks)
+
+    def swap_in(self,
+                sequence_group: SequenceGroup,
+                num_lookahead_slots: int = 0) -> Dict[int, int]:
+        blocks = self._get_blocks_for_swap(sequence_group,
+                                           SequenceStatus.SWAPPED,
+                                           num_lookahead_slots)
+        self.block_allocator.swap(blocks=blocks,
+                                  source_device=Device.CPU,
+                                  dest_device=Device.GPU)
+        # NOTE: Once the BlockManagerV1 implementation is deleted, we can
+        # move this get_and_reset_swaps call outside of swap_in/swap_out.
+        # Then the scheduler can make calls to get all swaps and all
+        # copy-on-writes for the batch.
+        mapping = self.block_allocator.get_and_reset_swaps()
+        block_number_mapping = {
+            cpu_block_id - self.num_total_gpu_blocks: gpu_block_id
+            for cpu_block_id, gpu_block_id in mapping.items()
+        }
+        return block_number_mapping
 
     def can_swap_out(self, seq_group: SequenceGroup) -> bool:
-        return False
-
-    def swap_out(self, seq_group: SequenceGroup) -> Dict[int, int]:
-        raise NotImplementedError
+        return self._can_swap(seq_group, Device.CPU, SequenceStatus.RUNNING)
+
+    def swap_out(self, sequence_group: SequenceGroup) -> Dict[int, int]:
+        blocks = self._get_blocks_for_swap(sequence_group,
+                                           SequenceStatus.RUNNING)
+        self.block_allocator.swap(blocks=blocks,
+                                  source_device=Device.GPU,
+                                  dest_device=Device.CPU)
+        mapping = self.block_allocator.get_and_reset_swaps()
+        block_number_mapping = {
+            gpu_block_id: cpu_block_id - self.num_total_gpu_blocks
+            for gpu_block_id, cpu_block_id in mapping.items()
+        }
+        return block_number_mapping
 
     def get_num_free_gpu_blocks(self) -> int:
         return self.block_allocator.get_num_free_blocks(Device.GPU)
 
     def get_num_free_cpu_blocks(self) -> int:
         return self.block_allocator.get_num_free_blocks(Device.CPU)
+
+    def _can_swap(self,
+                  seq_group: SequenceGroup,
+                  device: Device,
+                  status: SequenceStatus,
+                  num_lookahead_slots: int = 0,
+                  watermark_blocks: int = 0) -> bool:
+        blocks = self._get_blocks_for_swap(seq_group, status)
+        return self.block_allocator.can_swap(blocks, device,
+                                             num_lookahead_slots,
+                                             watermark_blocks)
```

Comment:
Two comments here:
1. `num_lookahead_slots` should not go down into the block allocator -- it is a BlockTable/BlockManager concept. The BlockTable should convert `num_lookahead_slots` into empty blocks, which can be passed into the block allocator `can_swap`. The reason why this is important (besides separating concerns) is the logic in can_swap inside the block allocator gets pretty complicated, hard to reason about edge cases like sliding window + lookahead.
2. watermark blocks should not be passed down either -- it is a concept limited to the BlockManager. Seems we can have the block allocator return a count of new blocks necessary to swap the sequeqnce, then we check watermark here?

Reply:
for 2, that's reasonable and possible. For 1, although it's reasonable, it's hard to achieve it because if we try to append dummy look ahead slots for a sequence and return the blocks, we need to introduce extra management logic outside `_get_blocks_for_swap()` to free those blocks. For sliding window, we can just use the `min(block_touched, sliding_window/block_size)` in block manager to compare with `free_block - water_mark`. WDYT?

Reply:
OK we can revisit (1) when we add sliding window.